text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,208 @@
+- Feature Name: `freeze_bytes`
+- Start Date: 2024-02-13
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- Feature Name: `freeze_uninit`
+
+# Summary
+[summary]: #summary
+
+Defines a language primitive for freezing uninitialized bytes, and library functions that use this language primitive.
+
+# Motivation
+[motivation]: #motivation
+
+In Rust, it is generally undefined behaviour to read uninitialized memory. Most types disallow uninitialized bytes as part of their validity invariant. 
+While in most cases, this is not a problem, as a temporarily held (or externally checked) uninitialized value can be stored as `MaybeUninit<T>`, in some rare cases it can be useful or desireable to handle uninitialized data as a "Don't Care" value, while still doing typed operations, such as arithmetic. 
+Freeze allows these limited Rust Programs to convert uninitialized data into useless-but-initialized bytes.
+
+Examples of uses:
+1. The major use for freeze is to read padding bytes of structs. This can be used for [zero-copy serialization](https://docs.rs/abomonation/latest/abomonation/) or a [generic wrapper arround standard atomic types](https://docs.rs/atomic/latest/atomic/struct.Atomic.html). 
+2. SIMD Code using masks can load a large value by freezing the bytes, doing lanewise arithmetic operations, then doing a masked store of the initialized elements. With additional primitives not specified here, this can allow for efficient partial load operations which prevent logical operations from going out of bounds (such a primitive could be defined to yield uninit for the lane, which could then be frozen).
+3. Low level libraries, such as software floating-point implementations, used to provide operations for compilers where uninit is considered a valid value for the provided operations.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Two new library interfaces are defined for unsafe code users:
+
+```rust
+// in module `core::ptr`
+pub unsafe fn read_freeze<T>(ptr: *const T) -> T;
+
+impl<T> *const T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+impl<T> *mut T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+
+// in module `core::mem`
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self;
+}
+```
```

Comment:
Poking around the library, the obvious parallel to this functionality is `zeroed`, and these methods don't really parallel that at all.

Namely, `zeroed` says: give me a *value* that is initialized with zero bytes. You still have to `assume_init` to assert that zeroing is valid for your type, but you do know what the bytes are.

To me, `freeze` here says: initialize a *memory location* by "freezing" its bytes. Again, you should still have to `assume_init` to assert that an arbitrary value is valid for your type, but you know that the bytes are intialized.

Another maybe-controversial decision is that I think that these methods should require `*mut T`, since we're *conceptually* writing over uninitialized memory and replacing it with an arbitrary value, even though no actual writing is occurring (and thus, no data races can occur). It also allows for things like miri to *actually* randomize the data in the buffers to help with testing, without worrying about making things unsound.

Maybe there could be some sort of "atomic freeze" that accepts an `Ordering`, but in general, it seems like you're probably going to want a mutable pointer here.

So, to that end, it feels more like the API should look like this instead:

```rust
// in core::ptr
pub fn freeze<T>(ptr: *mut T);

pub unsafe fn read_freeze<T>(ptr: *mut T) -> T {
    freeze(ptr);
    read(ptr)
}

impl<T> *mut T {
    pub fn freeze<T>(self);

    pub unsafe fn read_freeze<T>(self) -> T {
        self.freeze();
        self.read()
    }
}

impl<T> MaybeUninit<T> {
    pub fn freeze(&mut self);

    // and maybe:
    pub fn frozen() -> MaybeUninit<T> {
        let mut uninit = MaybeUninit::uninit();
        uninit.freeze();
        uninit
    }
}
```

Reply:
> To me, `freeze` here says: initialize a _memory location_ by "freezing" its bytes. Again, you should still have to `assume_init` to assert that an arbitrary value is valid for your type, but you know that the bytes are intialized.

I was confused about this as well, but reading the rest of the RFC I came to the conclusion that this is supposed to be a read-only operation which does nothing to the original memory location. It returns a "freezed version" of the read bytes but the original memory is left untouched.

Reply:
> Another maybe-controversial decision is that I think that these methods should require `*mut T`, since we're _conceptually_ writing over uninitialized memory and replacing it with an arbitrary value, even though no actual writing is occurring (and thus, no data races can occur). It also allows for things like miri to _actually_ randomize the data in the buffers to help with testing, without worrying about making things unsound.

No, this is actually literally a write to the target memory, with all of the consequences that follow from that. (No "conceptual" about it, there are cases where you need to actually issue a write instruction, fault on read-only pages, the whole nine yards.) The RFC proposes to use `*ptr = ptr.read_freeze()` as the way to express this operation.

Reply:
Right, although I guess the main controversy here is that it kind of forgets the reason why `MaybeUninit` exists: in general, it's very bad to pass around values when you're dealing with potentially uninitialized memory.

For example, take this simple use case: you want to read out a struct in full, *including* padding bytes, which are normally uninitialized. So, let's say we freeze-read that struct, copy it over a buffer of bytes, then write that out.

This is unsound. When you read the struct as a value, the padding is *still* uninitialized. What you actually want to do is cast the struct to bytes first, then freeze-read that into your buffer.

But again, this is still wrong. You need to cast to a slice of *maybe-uninit* bytes, since the padding is uninitialized, freeze-read that into a slice of *init* bytes, and write that.

This is why `fn zeroed<T>() -> T` is mostly discouraged; if you're working with anything other than integers, it's generally wrong to use. This is why I think that `fn read_freeze<T>(ptr: *mut T) -> T` is the wrong API; in most cases, the `T` input and the `T` output should be different, but here, they're the same, and you have to do a bunch of weird transmutation before you actually get to the thing you want to do.

If we treat `freeze` as actually working on the memory location, then the freezing and reading can be done separately, with all the necessary transmutes in between. Or, we have a `freeze_transmute_copy` type thing that freezes and does a `transmute_copy` afterward.

But ultimately, I feel like `freeze` returning a value is the wrong choice.

Reply:
> But again, this is still wrong. You need to cast to a slice of _maybe-uninit_ bytes, since the padding is uninitialized, freeze-read that into a slice of _init_ bytes, and write that.

I agree with the first part: if you read a struct `T` using `read_freeze::<T>(_) -> T`  then you will still not have any padding bytes. But `read_freeze::<[u8; size_of::<T>()]>()` should work and produce initialized bytes, including for the padding, because (according to the docs in the RFC) the freezing happens *before* the typed-read, so it should be safe to do as long as the target type is valid for every *initialized* bit pattern.

Using by-value instead of by-mut-ref freeze avoids a whole bunch of issues related to "what if we just elided the read?" on weird memory. (This has been discussed at length on [#t-opsem](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem).) It also makes it unusable on read-only memory. The API footgun concern is sound, but the by-value `MaybeUninit<T>` method should alleviate that concern, no?

Reply:
I think that the `MaybeUninit<T>` API is fine; like I said, either copying or mutating works as a general API and you can just Do The Right Thing with the internal operations to make it happen. My main concern is about the way that the raw-pointer methods require a very specific order of the steps involved to be sound, and that they make it too easy to effectively just do a normal read without freezing anything, but *thinking* you've done something else.

Reply:
> I'm not sure there's a "nice" API here that isn't project-safe-transmute

Can you say a bit more about this? I'm the head of PST, and would love gather thoughts of how `freeze` might impact ST APIs.

Would it make any sense for the intrinsic to have this signature:
```rust
pub fn freeze<T>(v: &T) -> [u8; size_of::<T>()];
```
I *think* this would avoid some of the safety footguns of the signature defined by the RFC, but I'm not sure if this limits its utility in any obvious (or non obvious) ways.

Reply:
safe-transmute (and existing user crates like bytemuck) could leverage the ability to check types that allow any initialized bitpattern to offer a safe version of `MaybeUninit::assume_init_freeze`, so it's not so much that project-safe-transmute would be affected by this, but it or a future library feature could offer nicer APIs compared to the primitives.

```rust
pub fn freeze<T>(v: &T) -> [u8; size_of::<T>()];
```

The problem is that this severely limits what you can actually do with the frozen value (no arithmetic, no complex structs). I'd prefer an `unsafe` api with wider usability than a safe one that limits what is possible to do with it. Also, that API in particular requires `generic_const_exprs`.

Reply:
```rust
pub fn freeze<T>(v: &T) -> [u8; size_of::<T>()];
```
As a safe fn as-written, this would also require `T: Freeze` (as in "`T` does not contain interior mutability") to prevent data races, (unless it is additionally specified that data-race-reads produce `undef` which is then frozen, or something, but that seems out of scope for this RFC).

Reply:
Indeed, I meant to write:
```rust
pub fn freeze<T>(v: &mut T) -> [u8; size_of::<T>()];
```

> Also, that API in particular requires `generic_const_exprs`.

Some considerations:
- `generic_const_exprs` is available within the Rust standard library
- `generic_const_exprs`, IIRC, is required to define such functions, not to *use* such functions. It very well might be possible to stabilize a such a `freeze` before stabilizing all or parts of `generic_const_exprs`
- Using this safe `freeze` as the core intrinsic does not preclude first stabilizing the `unsafe` facades proposed by the RFC (which would depend internally on this safe intrinsic, plus a transmute). This would reduce the scope of unsafety in the core library.
- Since we're discussing an API that is going to endure for decades after the stabilization of `generic_const_exprs`, we should seriously consider the API we'd *like* to see, not just the API that's currently possible now using only stable features of Rust.

Reply:
> `generic_const_exprs` is available within the Rust standard library

Only as an implementation detail. I was not aware that there was any permission to use it in a stably-exposed manner (which could not be removed if GCE gets removed or restricted for some reason).

Reply:
`generic_const_exprs` is considered incomplete and should *not* be used in the standard library.

Reply:
Also, the versions with the `Freeze` bound, or that take `&mut` cannot implement the read operation.

Reply:
There's no such general permission; that would have to be sought out if this core intrinsic were to be stabilized. Since an RFC is the place to discuss a vision of Rust's future, that's a possibility that should be entertained. Heck, it's even possible that some limited aspect of GCE is stabilized before the core intrinsic, in which case special permission is moot. Even if the core intrinsic is *never* stabilized with this signature, there's *still* value in having it be safe, so that at least the standard library can benefit from it being safe.

Reply:
As far as I know right now, t-types asks to not enable `generic_const_exprs` anywhere in the sysroot. That's based on me asking them a few months ago. All our incomplete features should be treated like that.

Reply:
I mostly agree with @clarfonthey and think that the fundamental intrinsic should look like `fn freeze<T>(ptr: *const T);`. Using such in-place API allows to make the API completely safe, which is a nice bonus.

The in-place API would also allow introduction of a similar intrinsic, let's call it `volatile_freeze`. It would act like an optimization barrier and be a more reliable alternative to [`black_box`](https://doc.rust-lang.org/core/hint/fn.black_box.html), which would be really useful for crates like [`zeroize`](https://docs.rs/zeroize).

As for `*mut T` vs `*const T`, does it even fundamentally matter? IIUC the difference in covariance/invariance does not matter here and everything else acts purely as a hint for programmers. There are valid cases when `freeze` may be used on `*const T` and casting it to `*mut T` would only add confusion, so I think it's fine for the intrinsic to accept `*const T`. 

I don't think the intrinsic should bother with the `[u8; size_of::<T>()]` stuff. It's better to leave it to higher-level APIs built on top of it. Plus, it's reasonable to expect that the intrinsic will be able to work with `*const [T]` pointers, which is obviously not compatible with the array-based proposals.

Reply:
`*const T` in signatures is not logically correct for an operation that mutates the destination. It makes it trivial to pass `&T` to it.

Reply:
And what exact soundness issues may it cause? `freeze` by itself should not generate any write instructions. Using it together with `assume_init` on a memory backed by `MADV_FREE`d pages should be considered unsound.

Reply:
First of all, even if it doesn't emit any hardware writes, it still has to do an AM write to replace the uninit bytes with init ones.

Second of all, that proposed safety constraint means its unsound to use on anything heap allocated.

Reply:
>it still has to do an AM write to replace the uninit bytes with init ones

I would not be so sure. I will not be surprised if it will be handled in a special way similarly to `undef`. And in early prototype stages it may be even emulated like this:
```rust
pub unsafe fn freeze<T>(p: *const T) {
    asm!(
        "# {}",
        in(reg) p,
        options(nostack),
    );
}
```

It's better to wait for someone like @RalfJung to draft how it may look at AM level.

>that proposed safety constraint means its unsound to use on anything heap allocated

This issue only applies to freed memory (which you should not use either way) and freshly allocated memory untouched by writes. I wouldn't call it "anything".

To `MADV_FREE`-proof you code `freeze` by itself is not sufficient. You would have to use `write_volatile` (or `write` + `freeze_volatile`). I would be **really** surprised to find that `freeze` introduces some kind of write instructions.

Reply:
I don't need Ralf Jung to say how it would work on the AM level, I can tell you that.

We can define freeze_in_place in terms of the proposed intrinsic
```rust
pub unsafe fn freeze_in_place<T>(p: *mut T){
    core::hint::assert_unchecked(p.is_aligned(align_of::<T>()));
    let ptr = p.cast::<[u8;size_of::<T>()]>();
    ptr.write(ptr.read_freeze())
}
```

Reply:
Note, the above does not work for slice DSTs, and the fact that it copies data makes unsuitable for zero-copy serialization. An in-place intrinsic without these limitations would be *very* useful (and a by-value function could be defined in terms of it). See https://github.com/rust-lang/rfcs/pull/3605#discussion_r1553630817 for additional discussion.

Reply:
>We can define freeze_in_place in terms of the proposed intrinsic

This looks... unwieldy, to say the least. Compare it with the straightforward alternative of doing:
```
pub unsafe fn read_freeze<T>(p: *const T) -> T {
    ptr::freeze(T);
    ptr::read(p)
}
```

And you haven't answered my questions about whether `freeze` will generate actual write instructions or not.

Reply:
I'd expect any in-place freeze operation to work on freshly heap-allocated uninit memory, so it must work on an MADV_FREE page.

Reply:
>it must work on an MADV_FREE page

You haven't explained how would it work without risk of overwriting something or requiring spooky magic. Would `freeze` need to generate many writes for types which cross page boundaries? What about systems with page sizes different from 4 KiB? Compiler now has to know about page sizes?

My opinion is that `freeze` by itself is a wrong tool for this. `MADV_FREE` is closer to shared memory or hardware registers, in the sense that it can change under your foot. You need an explicit write to "stabilize" such memory.

Reply:
Exactly, it requires explicit writes.
An in-place `freeze` would require doing an explicit write.

Reply:
> It's better to wait for someone like @RalfJung to draft how it may look at AM level.

@chorman0773 is right, these are write on the AM level as they change the AM state.

> You haven't explained how would it work without the risk of unintentionally overwriting something or requiring spooky magic.

He doesn't even want this operation so it's not up to him to explain that. It's up to the people proposing in-place freeze to explain how to make it practical.

"It's unsound on MADV_FREE memory" is *not* practical.

Reply:
>these are writes on the AM level as they change the AM state

Does it make the following code unsound?
```rust
#[repr(C)]
pub struct Padded {
    a: u8,
    b: u64,
    c: u8,
}

pub const SIZE: usize = core::mem::size_of::<Padded>();

impl Padded {
    pub fn zerocopy(&self) -> &[u8; SIZE] {
        let p = self as *const Self as *const [u8; SIZE];
        // Casting `*const T` to `*mut T` is fine per se:
        // https://users.rust-lang.org/t/55965
        ptr::freeze(p as *mut [u8; SIZE]);
        unsafe { &*p }
    }
}
```

If yes, then `freeze` looses a huge chunk of use cases desired in practice. If not, then why not make `freeze` accept `*const T` instead of making users to jump through hoops.

>It's up to the people proposing in-place freeze to explain how to make it practical.

In-place vs reading freeze is tangential to the `MADV_FREE` discussion. Well, assuming we want the reading `freeze` to allow useful optimizations like this:
```rust
fn foo(p: *mut [u8; 1 << 16]) {
    let buf: [u8; 1 << 16] = freeze(p);
    // `&mut buf` points to the same memory as `p`
    // instead of copying 64 KiB to stack
    do_stuff(&mut buf);
}
```

If such optimization is not allowed, then the proposed `freeze` becomes even more useless.

>"It's unsound on MADV_FREE memory" is not practical, in my understanding

My point is that `freeze` by itself is a *wrong* tool for this. In my opinion, `freeze` should be purely directive for compiler to drop `undef` optimizations for given allocation (regardless whether it's on stack, on heap, or somewhere else). For `MADV_FREE` we need a separate tool which would perform a small write into every page covered by referenced memory and perform freeze after that.

Reply:
> Does it make the following code unsound?

Yes. I have no strong opinion on whether in-place freeze should be provided -- it seems like a useful operation -- but if it is provided then it must be considered as potentially mutating all that memory.

It maybe *possible* to define an AM where a write that changes uninit to something else is not considered as violating the immutability of shared references. But that goes way beyond the scope of this RFC and is highly speculative. Note that currently even overwriting memory with itself (`*x = *x`) is considered as violating immutability, and AFAIK that is crucial -- such write *can* cause a data race and there are actual optimizations that break otherwise.

Reply:
You seem to be thinking of this is terms of permitted optimizations, but that's not a good way to specify a language -- it's non-modular and [error-prone](https://www.ralfj.de/blog/2020/12/14/provenance.html). Optimizations are a consequence of a given specification. For this example:
```rust
fn foo(p: *mut [u8; 1 << 16]) {
    let buf: [u8; 1 << 16] = freeze(p);
    // `&mut buf` points to the same memory as `p`
    // instead of copying 64 KiB to stack
    do_stuff(&mut buf);
}
```
That optimizations would be blatantly wrong. The code asked for a copy! The original code never mutates `p`. Calling `do_stuff` on `p` would obviously be a wrong optimization, and that doesn't even have anything to do with `freeze`.

With in-place freeze you could call that and then do stuff. With by-value freeze you have to do something like
```rust
fn foo(p: *mut [u8; 1 << 16]) {
  p.write(freeze(p.read());
  do_stuff(&mut buf);
}
```

Reply:
>You seem to be thinking of this is terms of permitted optimizations, but that's not a good way to specify a language

Yes, I know. But since I am more of a practitioner, it's easier for me to discuss concrete examples based on potential usecases I have in mind for `freeze`. Hopefully, then people like you who specialize in language semantics will be able to find a way to bridge what people want and how to specify it robustly.

Does not matter how robust your model is, if it does something surprising or unexpected for people who practice the language (be it it terms of performance or generated insructions), arguably, it's a bad model. You either need to better communicate what the model does, or change it.

After this discussion, I personally find that the proposed `freeze` model is hardly useful and that RFC did not do a good job of communicating capabilities of the model. `freeze` is usually viewed as an optimization technique and the proposed approach may result in very surprising (for low-level programmers) code generation. The RFC needs at the very least to clearly discuss several examples of what can be and can not be done with the model and provide expected code generation for each example. In the current form it does not even mention interaction with `MADV_FREE` like at all!

One last example from me:
```rust
pub fn f() {
    unsafe {
        let mut buf: [u8; 1024] = MaybeUninit::<[u8; 1024]>::uninit().freeze().assume_init();
        extern_read(&mut buf);
    }
}
```
How much stack space this function would it use when compiled with enabled optimizations? ~1024 bytes or ~2048 bytes? What exact writes would be generated?

Reply:
> How much stack space this function would use when compiled with enabled optimizations? ~1024 bytes? ~2048 bytes? More? What exact writes would be generated?

That example *should* be trivially optimizable to use 1kb of stack. No writes should be necessary since 1kb is less than a page size. If it was more than a page size then it's my understanding that stack probes would already be injected to probe the additional stack pages, irrespective of the use of `freeze` here.

Reply:
I wouldn't be so sure. Such buffer can easily cross page boundaries. And even if it does not, you don't know if any writes were generated for this page. What if `buf` was allocated right at the page edge? You may argue that for stack pages `MADV_FREE` is not used, but: 1) I can easily imagine threading library reusing stack frames with `MADV_FREE` 2) changing `freeze` behavior depending on whether it's used on stack or heap memory would be... not ideal, to say the least.

So instead of trivial no-op expected by programmers for this code snippet, we get quite non-trivial behavior.

Reply:
> So instead of trivial no-op expected by programmers for this code snippet, we get quite non-trivial behavior.

Yeah that's what happens when programmers are wrong. They expect things to be simple that aren't simple (see pointer provenance as another example of this). I don't blame them; optimizing compilers and system tricks like MADV_FREE make things complicated in subtle hard-to-predict ways. `freeze` needs careful documentation to make sure programmers are aware of the caveats.

Reply:
> What if buf was allocated right at the page edge?

Then you may need additional probes at the page boundary *if* the contract between the compiler and the platform does not require that stack reads give deterministic results already. I don't know if such a contract is already defined, but a reasonable option is to simply say that such a threading library is just wrong (in the same way the thread stacks are required to be properly aligned.)

> changing freeze behavior depending on whether it's used on stack or heap memory would be... not ideal, to say the least.

The behaviour would be exactly the same, because behaviour is a property of the AM. What instructions are emitted by the compiler for the intrinsic could be completely different in different circumstances though... Obviously? That's what compiler optimizations do!

> So instead of a trivial no-op expected by programmers for this code snippet, we get quite non-trivial behavior.

The behaviour is trivial, the emitted code is not. Freeze is not a no-op in the AM - whether it's a no-op in practice depends on optimizations and contracts, in the same way that whether a copy is elided depends on optimisations. If the contract says that stack memory reads are deterministic then it can indeed compile to a no-op. If not then of course the compiler needs to emit a write to that memory to ensure that it becomes initialized.